vizro-core/src/vizro/models/_models_utils.py

```diff
@@ -24,3 +25,53 @@ def get_unique_grid_component_ids(grid: List[List[int]]):
     unique_grid_idx = np.unique(grid)
     unique_grid_comp_idx = unique_grid_idx[unique_grid_idx != EMPTY_SPACE_CONST]
     return unique_grid_comp_idx
+
+
+# Validators for reuse
+def set_components(cls, components):
+    if not components:
+        raise ValueError("Ensure this value has at least 1 item.")
+    return components
+
+
+def set_layout(cls, layout, values):
+    from vizro.models import Layout
+
+    if "components" not in values:
+        return layout
+
+    if layout is None:
+        grid = [[i] for i in range(len(values["components"]))]
+        return Layout(grid=grid)
+
+    unique_grid_idx = get_unique_grid_component_ids(layout.grid)
+    if len(unique_grid_idx) != len(values["components"]):
+        raise ValueError("Number of page and grid components need to be the same.")
+    return layout
+
+
+def _create_component_container(self, components_content):
```

Comment:
I am wondering if this is the most efficient way to refactor these functions. These were previously private methods of the Page model, but they can now be re-used inside the Container and Form model. So it didn't make much sense to me to keep them as private methods to Page.

On the other hand, it's now more difficult to overwrite just these functions, as users now have to overwrite the entire build of the relevant model. Another option would be to wrap these functions inside private methods, but then we would have to duplicate them inside Page, Container, and Form.

Reply:
Would a potential solution be to have a 2 step process:
- `Container`, `Form` and `Page` have a method each to themselves
- that method just calls some commonly shared private fuction

This would keep the repetition to a minimum while allowing users to just overwrite their respective calling methods.

Reply:
Yeeees, that would be the last option I proposed. That would also be fine for me, but was wondering what people usually do here. Also wondering whether these functions are even something people would like to overwrite.

The whole assignment of grid areas could e.g. also be done via pure CSS so one wouldn't have to change these private methods.

Reply:
Returning to this one, I quickly wanted to show what I meant. I am convinced that the users don't need to overwrite these two private methods, `_create_component_container` and `_assign_component_grid_area.` And in this case, I don't see a strong reason anymore to keep them as private methods in the models. 

### Why is that?
I can only imagine people wanting to overwrite these private methods when they want to change our default layout, which is a CSS grid, to a Flex-Box layout. This can be the case if they don't care about a fixed grid layout but, e.g., want their components to be stacked subsequently (remember the prototype @antonymilne once built with the custom project cards). So there are two approaches:

#### 1. Use our existing CSS grid layout
No need to overwrite the private methods, as our default is working for them

**Result**
![Screenshot 2024-01-12 at 09 30 24](https://github.com/mckinsey/vizro/assets/90609403/f904004e-4b8e-451f-bbf7-326224b9665a)


#### 2. Use a Flexbox by leveraging custom CSS

```
.grid-layout {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: flex-start;
}

.card_container {
  min-width: 450px;
}
```

**Result**

![Screenshot 2024-01-12 at 09 45 49](https://github.com/mckinsey/vizro/assets/90609403/de8f1774-40fb-4cc1-ae85-2ea5bb3786d9)

@antonymilne - I realized method 2) would have saved you some time creating that layout 😅 

I think the above is essential to understand because leveraging custom CSS here is much easier than overwriting these private methods.

Reply:
Hm very interesting and you are probably right that the only time things would be changed by the user can be served by CSS. Would be happy to go with that solution then.

The only thing going for the other solution would then be the "we don't know what we don't know", as in who knows what ppl come up with, but that might quite unlikely :)

Reply:
Very good discussion 👍 You're absolutely right that CSS here would have been a much better way to do that layout with a variable number of cards.

Generally I agree with the approach here, but a few things I think we could improve further - let's discuss tomorrow:
1. the function signatures/separation of responsibility of the two helper functions don't feel quite right to me
2. `_models_utils` feels too generic as a home. These functions are specifically related to `Layout`, so , in order of smallest to biggest possible change:
  * maybe best to put them in `_layout` instead
  * in fact, why don't we even make them methods inside `Layout`? This makes it still possible to customise them by subclassing without any duplication
  * should we even have `Layout.build`? I did consider this in an early prototype and chose not to do it, but maybe it's a good idea now. wdyt? Let me see if I can remember why I didn't do it also.

Even if 99% of customisations that a user might ever want to do can be done through CSS, and we could consider the grid functionality as a fundamental part of Vizro that does not need to be user configurable, it would still be nice (but not very important) for this to be extensible through Python, which is most easily achieved by having the methods live in `Layout`.

Something else to consider is whether the same helper functions work already as they are in `Form`. Although maybe we ultimately won't need `Form` now that we have `Container` 🤔

Reply:
As discussed, we've moved the helper functions to the Layout model and introduced a build method there.

I went via 2 steps:
- Move everything to Layout and have `components` as an argument to the `Layout.build()`: https://github.com/mckinsey/vizro/pull/254/commits/a20859d6a2c18791c5f376a16bc206a094ff5776
- Remove components as an argument to the `Layout.build()` and call `component.build()` inside `Container, Page, Form`: https://github.com/mckinsey/vizro/pull/254/commits/591aae7ed930659511ee61775f47ebc3f3bfa4af


So the skeleton or the grid itself is created by the `Layout,` while the `components` are created and populated inside that skeleton within the relevant models 👍

Reply:
I think the second approach here is preferable overall. The only thing I didn't like is that relying on `children[idx].children` feels fragile, so I changed that to using a named key instead. See what you think: https://github.com/mckinsey/vizro/pull/273

Let me know if it makes sense. I haven't tested it at all so please do check before merging anything to `main` but hopefully it didn't break anything...